### PR TITLE
WI/MI CLI Phase 1 - Base API functionality

### PR DESCRIPTION
**Pull request by tsatam**
### Which issue this PR addresses:

Fixes [ARO-6453](https://issues.redhat.com/browse/ARO-6453)
Fixes [ARO-4398](https://issues.redhat.com/browse/ARO-4398)

### What this PR does / why we need it:

Implements functionality into the `az aro` extension to allow creating ARO clusters with a set of managed identities, instead of a service principal. 

As a consequence, this PR also includes changes to update the version of the `az aro` extension, as well as changes to upgrade the used ARO SDK/API to v20240812preview. 

### Test plan for issue:

- [X] Unit tests have been added for validation logic implemented across the three flags: `--enable-managed-identity, --assign-platform-workload-identity, --mi-user-assigned`
- [x] `az aro create` with the MIWI flags results in an API call being made to the RP with the neccessary fields populated
  - Example invocation:
    ```sh
    az aro create \
      --subscription ${SUBSCRIPTION_ID} \
      --resource-group ${CLUSTER_RESOURCE_GROUP} \
      --name azcli-testcluster \
      --version 4.14.16 
      --vnet aro-vnet --master-subnet master-subnet --worker-subnet worker-subnet \ 
      --enable-managed-identity \
      --assign-platform-workload-identity foo asdf \
      --assign-platform-workload-identity bar ghjkl \
      --mi-user-assigned baz
    ```
  - Produces request body:
    <details>
      <summary>Open</summary>
      
      ```json
      {
          "id": "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/${CLUSTER_RESOURCE_GROUP}/providers/Microsoft.RedHatOpenShift/openShiftClusters/azcli-testcluster",
          "location": "eastus",
          "identity": {
              "type": "UserAssigned",
              "userAssignedIdentities": {
                  "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/${CLUSTER_RESOURCE_GROUP}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/baz": {}
              }
          },
          "properties": {
              "clusterProfile": {
                  "pullSecret": "",
                  "domain": "fv0w9cvw",
                  "version": "4.14.16",
                  "resourceGroupId": "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/aro-fv0w9cvw",
                  "fipsValidatedModules": "Disabled"
              },
              "platformWorkloadIdentityProfile": {
                  "platformWorkloadIdentities": [
                      {
                          "operatorName": "foo",
                          "resourceId": "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/${CLUSTER_RESOURCE_GROUP}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/asdf"
                      },
                      {
                          "operatorName": "bar",
                          "resourceId": "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/${CLUSTER_RESOURCE_GROUP}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ghjkl"
                      }
                  ]
              },
              "networkProfile": {
                  "podCidr": "10.128.0.0/14",
                  "serviceCidr": "172.30.0.0/16",
                  "outboundType": "",
                  "preconfiguredNSG": "Disabled"
              },
              "masterProfile": {
                  "vmSize": "Standard_D8s_v3",
                  "subnetId": "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/${CLUSTER_RESOURCE_GROUP}/providers/Microsoft.Network/virtualNetworks/aro-vnet/subnets/master-subnet",
                  "encryptionAtHost": "Disabled"
              },
              "workerProfiles": [
                  {
                      "name": "worker",
                      "vmSize": "Standard_D2s_v3",
                      "diskSizeGB": 128,
                      "subnetId": "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/${CLUSTER_RESOURCE_GROUP}/providers/Microsoft.Network/virtualNetworks/aro-vnet/subnets/worker-subnet",
                      "count": 3,
                      "encryptionAtHost": "Disabled"
                  }
              ],
              "apiserverProfile": {
                  "visibility": "Public"
              },
              "ingressProfiles": [
                  {
                      "name": "default",
                      "visibility": "Public"
                  }
              ]
          }
      }
      ```
    </details>

### Is there any documentation that needs to be updated for this PR?

Not yet

### How do you know this will function as expected in production? 

Preview extension-only change. We will perform comprehensive testing before releasing this extension to users. 

### Additional notes
This PR is opened against ARO-RP's master branch. We may desire to isolate MIWI CLI changes to a dedicated branch, in order to afford us the ability to perform changes to our extension code pointing against the "stable" (v20231122) API, or to another future in-development API, but as we do not currently have any plans for parallel API efforts, I've chosen to merge these changes into our trunk branch.

### CODE REVIEW

File: python/az/aro/azext_aro/_validators.py
Comment (kimorris27) [inline python/az/aro/azext_aro/_validators.py:304]: The `if namespace.version is None` part of this conditional makes it so that the cx is going to have to explicitly pass a version even after the default install version is >= 4.14, but I don't think we have a way around enforcing this at the CLI level since the default install version is an "RP-side" thing - am I understanding correctly?

We could technically not enforce it at the CLI level and in doing so allow the request to hit the RP and fail from there, but I like this way you've done it so that a faulty request will fail as early as possible where possible.

File: python/az/aro/azext_aro/_validators.py
Comment (kimorris27) [inline python/az/aro/azext_aro/_validators.py:339]:
```suggestion
        raise RequiredArgumentMissingError('Must set --enable-managed-identity when providing a cluster identity')  # pylint: disable=line-too-long
```

Suggesting this change so that the message aligns with the parameter name.

File: python/az/aro/azext_aro/custom.py
Comment (kimorris27) [inline python/az/aro/azext_aro/custom.py:201]: Re: the comment on line 201, haven't you already added the cluster identity on line 198?

File: python/az/aro/azext_aro/tests/latest/unit/test_validators.py
Comment (kimorris27) [inline python/az/aro/azext_aro/tests/latest/unit/test_validators.py:985]:
```suggestion
        "Should raise InvalidArgumentValueError when version < 4.14.0",
```

Suggesting this change since 4.14.0 should be allowed.

File: python/az/aro/azext_aro/tests/latest/unit/test_validators.py
Comment (kimorris27) [inline python/az/aro/azext_aro/tests/latest/unit/test_validators.py:1033]: This list could use a `Should not raise any exception when valid` case where the identities are passed as resource IDs.

File: python/az/aro/azext_aro/tests/latest/unit/test_validators.py
Comment (kimorris27) [inline python/az/aro/azext_aro/tests/latest/unit/test_validators.py:1102]: This list could use a `Should not raise any exception when valid` case where the identity is passed as a resource ID.

**Review by kimorris27 — CHANGES_REQUESTED** [review]
LGTM overall! I suggested some small nitpicky changes.

File: python/az/aro/azext_aro/_validators.py
Comment (tsatam) [reply python/az/aro/azext_aro/_validators.py:304]: I think we'll need to push a dedicated CLI change after 4.14+ is the default, yeah. I don't see a better way around handling this since the CLI has no knowledge of what ARO's "default" install version is. 

I'm actually fine with removing our CLI-side validation if we want to just handle this in API static validation within the RP. I think that might be a better option.

File: python/az/aro/azext_aro/_validators.py
Comment (tsatam) [reply python/az/aro/azext_aro/_validators.py:339]: Good catch, thanks!

File: python/az/aro/azext_aro/custom.py
Comment (tsatam) [reply python/az/aro/azext_aro/custom.py:201]: Yes. I likely re-introduced the comment with a bad rebase. I'll remove it.

Comment (tsatam) [inline python/az/aro/azext_aro/custom.py:201]:
```suggestion
```

**Review by kimorris27 — APPROVED** [review]

**Review by kimorris27 — DISMISSED** [review]

File: python/az/aro/azext_aro/_actions.py
Comment (bitoku) [inline python/az/aro/azext_aro/_actions.py:12]: Was using validator not enough?
It looks to me that we can achieve the same with a validator and some formalizations after the parse.

I think we shouldn't do `protected-access` because it might be broken some day.

File: python/az/aro/azext_aro/_params.py
Comment (bitoku) [inline python/az/aro/azext_aro/_params.py:22]: nit: we can use `import (...)` just as `python/az/aro/azext_aro/tests/latest/unit/test_validators.py`

File: python/az/aro/azext_aro/_validators.py
Comment (bitoku) [inline python/az/aro/azext_aro/_validators.py:13]: nit: we can use `import (...)` just as `python/az/aro/azext_aro/tests/latest/unit/test_validators.py`

File: python/az/aro/azext_aro/_validators.py
Comment (bitoku) [inline python/az/aro/azext_aro/_validators.py:359]: nit: I like this because it starts with `is` and it is clearer that it returns bool.
```suggestion
def is_valid_identity_resource_id(rid):
```

File: python/az/aro/azext_aro/_validators.py
Comment (bitoku) [inline python/az/aro/azext_aro/_validators.py:327]: It seems it does more than validation.
I think it would be better that we stick to validation in validate functions, and if we need formalization, do somewhere else, otherwise it will be confusing.

btw is it safe to assume that `identity.resource_id` is a MI name when it is not valid?

File: python/az/aro/azext_aro/_validators.py
Comment (bitoku) [inline python/az/aro/azext_aro/_validators.py:342]: It seems it does more than validation.
I think it would be better that we stick to validation in validate functions, and if we need formalization, do somewhere else, otherwise it will be confusing.

btw is it safe to assume that `identity.mi_user_assigned` is a MI name when it is not valid?

File: python/az/aro/azext_aro/tests/latest/unit/test_validators.py
Comment (bitoku) [inline python/az/aro/azext_aro/tests/latest/unit/test_validators.py:6]: nit: I would make them one by one line. This would make diff cleaner.

```suggestion
    validate_cidr,
    validate_client_id,
    validate_client_secret,
```

**Review by bitoku — COMMENTED** [review]
Sorry for a lot of nits.
LGTM for the business logic itself, but I'd like you to take a look at non-nit reviews.

File: python/az/aro/azext_aro/_actions.py
Comment (tsatam) [reply python/az/aro/azext_aro/_actions.py:12]: This implementation follows the Azure CLI recommendations for implementing "action"-based parameters like `--assign-platform-workload-identity`: https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md#collection-management-with-actions 

It might be possible to do this with a custom implementation, but given that this code will eventually be upstreamed to the Azure CLI I'd prefer to just follow their conventions.

File: python/az/aro/azext_aro/_validators.py
Comment (tsatam) [reply python/az/aro/azext_aro/_validators.py:327]: I did this to keep in line with our "validation" performed on vnets/subnets (that does the same name -> resourceid conversion): https://github.com/Azure/ARO-RP/blob/tsatam/ARO-4398-wimi-cli-p1-base-api-functionality/python/az/aro/azext_aro/_validators.py#L239

I agree that this is confusing, it tripped me up when implementing this. Perhaps we move these conversions to the Actions implementation used above? 

This is something I'd like to tackle in an overall refactor, but perhaps not within this PR itself.

Comment (tsatam) [reply python/az/aro/azext_aro/_validators.py:327]: As for our name / resource ID handling here, we are not yet validating that the identity resource even exists yet. 

Ultimately this will be handled by API-side validation, and we will introduce checks for the identity existing further along in MIWI efforts, once we introduce validation of the passed-in identities and their roleassignments.

File: python/az/aro/azext_aro/_validators.py
Comment (tsatam) [reply python/az/aro/azext_aro/_validators.py:342]: Same response as https://github.com/Azure/ARO-RP/pull/3624#discussion_r1657599264

File: python/az/aro/azext_aro/tests/latest/unit/test_validators.py
Comment (tsatam) [reply python/az/aro/azext_aro/tests/latest/unit/test_validators.py]: I made this change globally across our imports for consistency's sake.

File: python/az/aro/azext_aro/_params.py
Comment (tsatam) [reply python/az/aro/azext_aro/_params.py]: I made this change globally across our imports for consistency's sake.

File: python/az/aro/azext_aro/_validators.py
Comment (tsatam) [reply python/az/aro/azext_aro/_validators.py]: I made this change globally across our imports for consistency's sake.

File: python/az/aro/azext_aro/_validators.py
Comment (tsatam) [reply python/az/aro/azext_aro/_validators.py]: Done

File: python/az/aro/azext_aro/_actions.py
Comment (bitoku) [reply python/az/aro/azext_aro/_actions.py:12]: I didn't know there's a doc about it. thanks!

File: python/az/aro/azext_aro/_validators.py
Comment (bitoku) [reply python/az/aro/azext_aro/_validators.py:330]: thanks! fair enough

> Perhaps we move these conversions to the Actions implementation used above?
It would be better I guess.

I was worried about `identity.resource_id` conversion because if there's just a typo, it could be interpreted as a resource name.
However, on second thoughts, it doesn't create anything, and even though there's a typo, it'll end up just getting an error, so it isn't a big deal.

**Review by bitoku — DISMISSED** [review]

**Review by SudoBrendan — APPROVED** [review]
This all seems OK to me - if you want to move some ocp version validation to RP frontend that would be fine but I wouldn't call it required.

**Review by kimorris27 — DISMISSED** [review]

**Review by kimorris27 — APPROVED** [review]
The current E2E failure is an issue with the pipeline and not with these code changes.

File: python/az/aro/azext_aro/custom.py
Comment (SudoBrendan) [inline python/az/aro/azext_aro/custom.py:215]: q: is this required for this PR, or something for the future?

**Review by SudoBrendan — APPROVED** [review]
Seems fine to me, but to be sure - is the TODO required before merge?

File: python/az/aro/azext_aro/custom.py
Comment (tsatam) [reply python/az/aro/azext_aro/custom.py:215]: It's for the future.

**Comment by tsatam** [comment]
Merging this PR now to unblock #3709

**Pull request merged**
